File: js/perf_log.js
/**
 * Env-gated performance logging for renderer (page) code.
 *
 * Mirrors py/perf_log.py. Emits "LOG: perf <label> <ms>" lines which are sent
 * over IPC to the main process ("perfLog" channel) so they flow through the
 * existing console.log pipeline into masonjar.log and the in-app Log window,
 * exactly like Python LOG lines.
 *
 * Enable by launching with the environment variable MASONJAR_PERF=1 (also
 * accepts true/yes/on). No-op when disabled, so it is safe to leave in place.
 *
 * Usage (in a page script):
 *     var perfLog = require("./perf_log");
 *     perfLog.perfLog("page.foo.step", ms);
 *     var t0 = perfLog.now(); ...; perfLog.perfLog("label", perfLog.now() - t0);
 */

var PERF_ENABLED = false;
try {
	var _v = (process && process.env && process.env.MASONJAR_PERF) || "";
	PERF_ENABLED =
		["1", "true", "yes", "on"].indexOf(String(_v).trim().toLowerCase()) >= 0;
} catch (_e) {
	PERF_ENABLED = false;
}

var _ipc = null;
if (PERF_ENABLED) {
	try {
		_ipc = require("electron").ipcRenderer;
	} catch (_e2) {
		_ipc = null;
	}
}

function perfEnabled() {
	return PERF_ENABLED;
}

function _emit(line) {
	try {
		if (_ipc) {
			_ipc.send("perfLog", line);
		} else if (typeof console !== "undefined") {
			// Fallback (e.g. no ipcRenderer): at least reach devtools console.
			console.log(line);
		}
	} catch (_e) {
		// ignore
	}
}

function now() {
	try {
		return performance.now();
	} catch (_e) {
		return Date.now();
	}
}

function perfLog(label, ms) {
	if (!PERF_ENABLED) {
		return;
	}
	_emit("LOG: perf " + label + " " + Number(ms).toFixed(1) + "ms");
}

function perfSection(label, fn) {
	if (!PERF_ENABLED) {
		return fn();
	}
	var t0 = now();
	try {
		return fn();
	} finally {
		perfLog(label, now() - t0);
	}
}

function pageName() {
	try {
		var p = (window.location && window.location.pathname) || "";
		var base = p.split("/").pop() || "page";
		return base.replace(/\.html$/i, "") || "page";
	} catch (_e) {
		return "page";
	}
}

var _pageLoadEmitted = false;
function emitPageLoadOnce() {
	if (!PERF_ENABLED || _pageLoadEmitted) {
		return;
	}
	_pageLoadEmitted = true;
	var emit = function () {
		try {
			var entries =
				(performance.getEntriesByType &&
					performance.getEntriesByType("navigation")) ||
				[];
			var ms = entries[0] ? entries[0].duration : now();
			perfLog("page." + pageName() + ".load", ms);
		} catch (_e) {
			// ignore
		}
	};
	try {
		if (typeof document !== "undefined" && document.readyState === "complete") {
			emit();
		} else if (typeof window !== "undefined") {
			window.addEventListener("load", emit, { once: true });
		}
	} catch (_e) {
		emit();
	}
}

// Auto-run page-load timing whenever this module is loaded in a renderer page.
try {
	emitPageLoadOnce();
} catch (_e) {
	// ignore
}

module.exports = {
	perfEnabled: perfEnabled,
	perfLog: perfLog,
	perfSection: perfSection,
	now: now,
	pageName: pageName,
	emitPageLoadOnce: emitPageLoadOnce,
};

File: py/perf_log.py
"""Lightweight, env-gated performance timing for Mason Jar pipeline steps.

Enable by setting the environment variable ``MASONJAR_PERF=1`` (also accepts
true/yes/on). All output is written to stdout as ``LOG: perf <label> <ms>``
lines, which the app already captures into ``~/.masonjar/masonjar.log`` and the
in-app Log window. When disabled the helpers are near-zero cost (no timing, no
output), so they are safe to leave in production code.

Usage:
    import perf_log

    # Total wall time of a whole step (one line at the entry point):
    perf_log.perf_start_total("count")

    # Time a specific block:
    with perf_log.perf_section("count.load_masks"):
        ...

    # Decorate a function:
    @perf_log.perf_timed("count.per_slice")
    def process(...):
        ...

Review: set MASONJAR_PERF=1, run the step, then filter ``LOG: perf`` in
masonjar.log or the in-app Log window.
"""

from __future__ import annotations

import atexit
import os
import time
import tracemalloc
from contextlib import contextmanager

PERF_ENABLED = os.environ.get("MASONJAR_PERF", "").strip().lower() in (
    "1",
    "true",
    "yes",
    "on",
)

if PERF_ENABLED:
    # Python allocation peak is portable and has near-zero overhead when the
    # existing performance switch is off. RSS is added when psutil is present.
    tracemalloc.start()


def perf_enabled() -> bool:
    """Return True when perf logging is enabled via MASONJAR_PERF."""
    return PERF_ENABLED


def perf_log(label: str, ms: float) -> None:
    """Emit a single ``LOG: perf <label> <ms>`` line (only when enabled)."""
    if PERF_ENABLED:
        print(f"LOG: perf {label} {ms:.1f}ms", flush=True)


def perf_memory(label: str) -> None:
    """Emit traced Python peak and optional process RSS in MiB."""
    if not PERF_ENABLED:
        return
    _current, peak = tracemalloc.get_traced_memory()
    rss = None
    try:
        import psutil  # type: ignore

        rss = psutil.Process().memory_info().rss / 1048576.0
    except Exception:
        pass
    suffix = f" rss={rss:.1f}MiB" if rss is not None else ""
    print(f"LOG: memory {label} python_peak={peak / 1048576.0:.1f}MiB{suffix}", flush=True)


@contextmanager
def perf_section(label: str):
    """Time a code block; emits the duration on exit when enabled.

    Near-zero cost when disabled (no perf_counter calls, no output).
    """
    if not PERF_ENABLED:
        yield
        return
    _t = time.perf_counter()
    try:
        yield
    finally:
        perf_log(label, (time.perf_counter() - _t) * 1000.0)


def perf_timed(label: str):
    """Decorator variant of :func:`perf_section`."""

    def _decorator(fn):
        def _wrapped(*args, **kwargs):
            if not PERF_ENABLED:
                return fn(*args, **kwargs)
            _t = time.perf_counter()
            try:
                return fn(*args, **kwargs)
            finally:
                perf_log(label, (time.perf_counter() - _t) * 1000.0)

        return _wrapped

    return _decorator


def perf_start_total(label: str) -> None:
    """Record process start and emit total wall time on interpreter exit.

    A single call at a step's entry point instruments the whole step without
    wrapping or re-indenting its body. Each pipeline step runs as its own
    Python process, so the atexit handler fires exactly once per step.
    """
    if not PERF_ENABLED:
        return
    _t0 = time.perf_counter()
    atexit.register(
        lambda: print(
            f"LOG: perf {label}.total "
            f"{(time.perf_counter() - _t0) * 1000.0:.1f}ms",
            flush=True,
        )
    )
